### ISSUE TITLE
Training error. Help

### ISSUE THREAD

**Issue by ilichev-andrey** [torch/torch7#856]
Hello.
I teach a neural network for two of my classes.
Error occurs at the stage of training.
How to fix it?

I'm new to this. Help me please.

th> require 'nn';
th> trainset = torch.load('animals_peoples2.t7')
th> testset = torch.load('animals_peoples2.t7')
th> classes = {'animals', 'peoples'}

th> print(trainset)
{
data : ByteTensor - size: 17299x3x96x96
label : ByteTensor - size: 17299
}

th> print(#trainset.data)
17299
3
96
96
[torch.LongStorage of size 4]

th> setmetatable(trainset,
..> {__index = function(t, i)
..> return {
..> t.data[i],
..> t.label[i]
..> }
..> end}
..> );

th> function trainset:size()
..> return self.data:size(1)
..> end

th> trainset.data = trainset.data:double()

th> print(trainset:size())
17299

th> print(trainset[33])
{
1 : DoubleTensor - size: 3x96x96
2 : 1
}

th> redChannel = trainset.data:select(2, 1)

th> print(#redChannel)
17299
96
96
[torch.LongStorage of size 3]

th> mean = {} -- store the mean, to normalize the test set in the future

th> stdv = {} -- store the standard-deviation for the future

th> for i=1,3 do -- over each image channel
..> mean[i] = trainset.data:select(2, 1):mean() -- mean estimation
..> print('Channel ' .. i .. ', Mean: ' .. mean[i])
..> trainset.data:select(2, 1):add(-mean[i]) -- mean subtraction
..>
..> stdv[i] = trainset.data:select(2, i):std() -- std estimation
..> print('Channel ' .. i .. ', Standard Deviation: ' .. stdv[i])
..> trainset.data:select(2, i):div(stdv[i]) -- std scaling
..> end
Channel 1, Mean: 0
Channel 1, Standard Deviation: 0
Channel 2, Mean: nan
Channel 2, Standard Deviation: 0
Channel 3, Mean: nan
Channel 3, Standard Deviation: 0

th> net = nn.Sequential()
th> net:add(nn.SpatialConvolution(3, 6, 18, 18)) -- 3 input image channels, 6 output channels, 18x18 convolution kernel
th> net:add(nn.ReLU()) -- non-linearity
th> net:add(nn.SpatialMaxPooling(2,2,2,2)) -- A max-pooling operation that looks at 2x2 windows and finds the max.
th> net:add(nn.SpatialConvolution(6, 16, 18, 18))
th> net:add(nn.ReLU()) -- non-linearity
th> net:add(nn.SpatialMaxPooling(2,2,2,2))
th> net:add(nn.View(16x18x18)) -- reshapes from a 3D tensor of 16x18x18 into 1D tensor of 16x18x18
th> net:add(nn.Linear(16x18x18, 120)) -- fully connected layer (matrix multiplication between input and weights)
th> net:add(nn.ReLU()) -- non-linearity
th> net:add(nn.Linear(120, 84))
th> net:add(nn.ReLU()) -- non-linearity
th> net:add(nn.Linear(84, 2))
th> net:add(nn.LogSoftMax()) -- converts the output to a log-probability. Useful for classification problems

nn.Sequential {
[input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> (13) -> output]
(1): nn.SpatialConvolution(3 -> 6, 18x18)
(2): nn.ReLU
(3): nn.SpatialMaxPooling(2x2, 2,2)
(4): nn.SpatialConvolution(6 -> 16, 18x18)
(5): nn.ReLU
(6): nn.SpatialMaxPooling(2x2, 2,2)
(7): nn.View(5184)
(8): nn.Linear(5184 -> 120)
(9): nn.ReLU
(10): nn.Linear(120 -> 84)
(11): nn.ReLU
(12): nn.Linear(84 -> 2)
(13): nn.LogSoftMax
}
[0.0371s]
th> criterion = nn.ClassNLLCriterion()
[0.0761s]
th>
[0.0000s]
th> trainer = nn.StochasticGradient(net, criterion)
[0.0001s]
th> trainer.learningRate = 0.001
[0.0000s]
th> trainer.maxIteration = 5
[0.0000s]
th> trainer:train(trainset)
StochasticGradient: training

.../facedetect/torch/install/share/lua/5.1/nn/Container.lua:67:
In 7 module of nn.Sequential:
/root/facedetect/torch/install/share/lua/5.1/nn/View.lua:47: input view (16x11x11) and desired view (5184) do not match
stack traceback:
[C]: in function 'error'
/root/facedetect/torch/install/share/lua/5.1/nn/View.lua:47: in function 'batchsize'
/root/facedetect/torch/install/share/lua/5.1/nn/View.lua:79: in function </root/facedetect/torch/install/share/lua/5.1/nn/View.lua:77>
[C]: in function 'xpcall'
.../facedetect/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
...facedetect/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
...ct/torch/install/share/lua/5.1/nn/StochasticGradient.lua:35: in function 'train'
[string "_RESULT={trainer:train(trainset)}"]:1: in main chunk
[C]: in function 'xpcall'
/root/facedetect/torch/install/share/lua/5.1/trepl/init.lua:661: in function 'repl'
...tect/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
[C]: at 0x004064f0

WARNING: If you see a stack trace below, it doesn't point to the place where this error occurred. Please use only the one above.
stack traceback:
[C]: in function 'error'
.../facedetect/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'
...facedetect/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
...ct/torch/install/share/lua/5.1/nn/StochasticGradient.lua:35: in function 'train'
[string "_RESULT={trainer:train(trainset)}"]:1: in main chunk
[C]: in function 'xpcall'
/root/facedetect/torch/install/share/lua/5.1/trepl/init.lua:661: in function 'repl'
...tect/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:199: in main chunk
[C]: at 0x004064f0
[0.3159s]

**Comment by ilichev-andrey** [comment]
how to determine these parameters:

5x5 convolution kernel?
net:add(nn.SpatialConvolution(3, 6, 5, 5))

reshapes from a 3D tensor of 16x5x5 into 1D tensor of 16*5*5?
net:add(nn.View(16*5*5))

**Comment by fmassa** [comment]
After all the convolutions and max poolings, your input has size 16x11x11, but your layer was expecting 16x18x18. You should fix that.
Also, the issues are for tracking bugs. In the future, consider using the gitter or the google groups for questions
